### PR TITLE
Feature/add error handler for callback

### LINKED ISSUE #32: Subscribe to feature refresh errors
I recently created a #31  to include a callback on error feature. I want an answer about this. I am ready to correct my pull request and add tests. Thank you)

### PR DESCRIPTION
**Pull request by AmirSalakhov94**
Hello. We want to use client-side Growthbook service call metrics. But a handler made for succesfully calls only would like to add an error handler as well. Here's the pull request, it's not claims to be the best solution. Thank you)

Closes #32

### CODE REVIEW

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java [hunk 1]
```diff
@@ -98,9 +98,8 @@ public String getFeaturesJson() {
 
     /**
      * Subscribe to feature refresh events
-     * This callback is called when the features are successfully refreshed.
+     * This callback is called when the features are successfully or error refreshed.
```
Comment (tinahollygb) [inline lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java:101]:
```suggestion
     * This callback is called when the features are successfully refreshed or there is an error when refreshing.
```

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java [hunk 2]
```diff
@@ -246,7 +245,15 @@ private void onSuccess(Response response) throws FeatureFetchException {
         }
     }
 
+    private void refreshSuccessfullyCallbacks(String featuresJson) {
```
Comment (tinahollygb) [inline lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java:248]: Can you rename this to `onRefreshSuccess(String featuresJson)`

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java [hunk 3]
```diff
@@ -246,7 +245,15 @@ private void onSuccess(Response response) throws FeatureFetchException {
...
+        this.refreshCallbacks
+                .forEach(featureRefreshCallback -> featureRefreshCallback.onRefresh(featuresJson));
+    }
 
+    private void refreshFailedCallbacks(Throwable throwable) {
```
Comment (tinahollygb) [inline lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java:253]: Can you rename this to `onRefreshFailed(Throwable throwable)`

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java [hunk 4]
```diff
@@ -233,9 +233,8 @@ private void onSuccess(Response response) throws FeatureFetchException {
 
             this.featuresJson = refreshedFeatures;
 
-            this.refreshCallbacks.forEach(featureRefreshCallback -> {
-                featureRefreshCallback.onRefresh(this.featuresJson);
-            });
+            this.refreshSuccessfullyCallbacks(this.featuresJson);
+
```
Comment (tinahollygb) [inline lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java:237]: Style: Unnecessary whitespace

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java [hunk 5]
```diff
@@ -246,7 +245,15 @@ private void onSuccess(Response response) throws FeatureFetchException {
         }
     }
 
+    private void refreshSuccessfullyCallbacks(String featuresJson) {
+        this.refreshCallbacks
+                .forEach(featureRefreshCallback -> featureRefreshCallback.onRefresh(featuresJson));
```
Comment (tinahollygb) [inline lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java:250]: Style: Can you use the existing style (add braces).

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java [hunk 6]
```diff
@@ -246,7 +245,15 @@ private void onSuccess(Response response) throws FeatureFetchException {
...
+    }
 
+    private void refreshFailedCallbacks(Throwable throwable) {
+        this.refreshCallbacks
+                .forEach(featureRefreshCallback -> featureRefreshCallback.onError(throwable));
```
Comment (tinahollygb) [inline lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java:255]: Same style request here as above re: braces

**Review by tinahollygb — CHANGES_REQUESTED** [review]
Functionality-wise seems good, but I left some grammar and style requests. See inline. Thanks!

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java (hunk 4 above)
Comment (AmirSalakhov94) [reply lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java]: corrected

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java (hunk 6 above)
Comment (AmirSalakhov94) [reply lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java]: corrected

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java (hunk 5 above)
Comment (AmirSalakhov94) [reply lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java]: corrected)

File: lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java (hunk 3 above)
Comment (AmirSalakhov94) [reply lib/src/main/java/growthbook/sdk/java/GBFeaturesRepository.java]: corrected)

**Comment by AmirSalakhov94** [comment]
> Functionality-wise seems good, but I left some grammar and style requests. See inline. Thanks!

Thank you, I was glad to contribute) I added 2 more tests to check refresh callback.

File: lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java [hunk 7]
```diff
@@ -137,8 +139,65 @@ void canFetchEncryptedFeatures_mockedResponse() throws IOException, FeatureFetch
...
+            public void onRefresh(String featuresJson) {
+            }
 
+            @Override
+            public void onError(Throwable throwable) {
+                assertThrows(IOException.class, () -> {
+                    throw throwable;
+                });
+            }
+        });
```
Comment (tinahollygb) [inline lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java:197]: I don't think this assertion is working as expected. I'm assuming you're trying to test that the onError callback is called when it fails, but when I comment out the real code, this test still passes, which leads me to believe it isn't testing anything.

Comment (AmirSalakhov94) [reply lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java]: I slightly corrected the test, if I understood you correctly, now when you comment on the real code, it ends with an error. If you have any ideas how it can be implemented, I will be grateful)

Comment (tinahollygb) [reply lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java]: There is some strange behaviour with the tests still. I notice that the `getResponse` method you've abstracted from the initial success is used for the failure, and it responds with a 200. I see that when I change the failed response JSON, it fails because it tries to successfully process the failed JSON response data. It looks like both the success and failure cases are being called in the tests. I don't think the failure tests are accurately being tested—we are testing a condition that would never occur (both a success and a failure). You can see the odd behaviour by changing `testOnFeaturesRefresh_Error()`'s fake response to be `"{}"`—we get: `CONFIGURATION_ERROR : No features found` which means it's trying to run the success.

**Review by tinahollygb — CHANGES_REQUESTED** [review]
See inline comment but tests are still not accurately testing a real failure case scenario.

File: lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java (hunk 7 above)
Comment (AmirSalakhov94) [reply lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java]: oh yeah i see...it's happening because i'm calling the subject.initialize() method, it's not really related to the error check as it doesn't call onRefreshFailed, i can remove it from the call and leave only subject.getFeaturesJson(), then the test will work correctly

File: lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java [hunk 8]
```diff
@@ -158,19 +206,54 @@ private static OkHttpClient mockHttpClient(final String serializedBody) throws I
...
+    /**
+     * Create a mock instance for failed callback of {@link OkHttpClient}
+     *
+     * @param serializedBody JSON string response
+     * @return mock {@link OkHttpClient}
+     */
+    private static OkHttpClient mockHttpClientFailedCallback(final String serializedBody, final IOException ioException) throws IOException {
+        OkHttpClient okHttpClient = mock(OkHttpClient.class);
+        Call remoteCall = mock(Call.class);
+        when(remoteCall.execute()).thenReturn(getResponse(serializedBody));
```
Comment (tinahollygb) [inline lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java:230]: It might make sense to pass the status code into `getResponse`—it would make more sense to say that a failure would return an error code instead of 200.

Alternatively, don't abstract `getResponse` (that'll probably make it easier to read) and just do a new failure response inline where an error status is used (and not 200).

Comment (AmirSalakhov94) [reply lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java]: @tinahollygb hi) so, I rolled back my changes in the test that affects your implementation. Аdded for test testOnFeaturesRefresh_Error failed response.

File: lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java [hunk 9]
```diff
@@ -159,14 +234,14 @@ private static OkHttpClient mockHttpClient(final String serializedBody) throws I
...
-                ))
-            .build();
+                .request(new Request.Builder().url("http://url.com").build())
+                .protocol(Protocol.HTTP_1_1)
+                .code(200).message("").body(
+                        ResponseBody.create(
+                                serializedBody,
+                                MediaType.parse("application/json")
+                        ))
+                .build();
```
Comment (tinahollygb) [inline lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java:244]: Can you clean up these unnecessary changes?

Comment (AmirSalakhov94) [reply lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java]: sorry it's because of my auto style code

**Review by tinahollygb — APPROVED** [review]

**Comment by tinahollygb** [comment]
Awesome, thanks for the contribution! 🚀

**Comment by AmirSalakhov94** [comment]
> Awesome, thanks for the contribution! 🚀

Glad to take part)
